### DISCUSSION TITLE
Help writing a plugin

### DISCUSSION THREAD

**Discussion by darko1979** [Novik/ruTorrent#2294] [Q&A] [answered]
I had an issue multiple times where I would lose values in 'Added' column, or it would be overwritten after a forced check.
When this happens I would preferer to have file modified time here.
I was looking at the source code of the plugins and wrote something like this:
```
if(plugin.canChangeMenu())
{
  plugin.createMenu = theWebUI.createMenu;
  theWebUI.createMenu = function( e, id )
  {
    plugin.createMenu.call(this, e, id);
    if(plugin.enabled && id && id.length == 40)
    {
      plugin.uID = id;
      theContextMenu.add( ["Fix Added Time", "theWebUI.perform('fixAddTime')"] );
    }
  }
  rTorrentStub.prototype.fixAddTime = function() {
    alert(`fixAddTimeStub: ${plugin.uID}`);

    if (plugin.uID && plugin.uID.length == 40) {
      var cmd = new rXMLRPCCommand("d.set_custom");
      cmd.addParameter("string", plugin.uID);
      cmd.addParameter("string", "addtime");
      cmd.addParameter("string", "1644316034");
      // cmd.addParameter("string", "(cat,$d.creation_date)");
      this.commands.push(cmd);
    }
  }
}
```
I need help with this line:
```
cmd.addParameter("string", "1644316034");
```
I need to get modified time from one of the files of the selected torrent but I'm not sure how to get it.

**Comment by darko1979** [comment, marked as answer]
Ok, I managed to accomplish what I wanted.
I forgot to tell that I wrote this plugin for rTorrent 0.9.6 and ruTorrent 0.13.6 (this is what I get from seedbox).
Here is the code for init.js
```
if(plugin.canChangeMenu())
{
  plugin.createMenu = theWebUI.createMenu;
  theWebUI.createMenu = function( e, id )
  {
    plugin.createMenu.call(this, e, id);
    if(plugin.enabled && id && id.length == 40)
    {
      plugin.uID = id;
      theContextMenu.add( ["Fix Added Time", "theWebUI.perform('fix_addtime_request_action')"] );
    }
  }

  rTorrentStub.prototype.fix_addtime_request_action = function() {
    // console.log(`fix_addtime_request_action: ${plugin.uID}`);
    if (plugin.uID && plugin.uID.length == 40) {
      theWebUI.requestWithoutTimeout("?action=fix_addtime_action",[plugin.onDataReceived, plugin]); 
    }
  }

  rTorrentStub.prototype.fix_addtime_action = function() {
    // console.log(`fix_addtime_action(hash: ${plugin.uID})`);
    if (plugin.uID && plugin.uID.length == 40) {
      var base_path = theWebUI.torrents[plugin.uID].base_path;
      this.content = "path=" + base_path;
      this.contentType = "application/x-www-form-urlencoded";
      this.mountPoint = "plugins/fix_addtime/action.php";
      this.dataType = "json";
    }
  }

  plugin.onDataReceived = function (data)
  {
    // console.log(`onDataReceived(hash: ${plugin.uID}  data: ${JSON.stringify(data)})`);
    if (plugin.uID && plugin.uID.length == 40 && data && data.mtime) {
      plugin.mtime = data.mtime;
      theWebUI.requestWithoutTimeout("?action=fix_addtime_set",[plugin.onDataReceived, plugin]); 
    }
   }

  rTorrentStub.prototype.fix_addtime_set = function() {
    // console.log(`fix_addtime_set(hash: ${plugin.uID}, mtime: ${plugin.mtime})`);
    if (plugin.uID && plugin.uID.length == 40 && plugin.mtime) {
      // console.log(`updating addtime with ${plugin.mtime}`);
      try {
        var cmd = new rXMLRPCCommand("d.set_custom");
        cmd.addParameter("string", plugin.uID);
        cmd.addParameter("string", "addtime");
        cmd.addParameter("string", `${plugin.mtime}`);
        this.commands.push(cmd);
        // console.log('done.');
      } catch (err) {
        console.log(err);
      }
    }
  }
}
```
And here is the code for action.php
```
<?php
require_once( '../../php/util.php' );
$result = array();
if (isset($_REQUEST['path'])) {
  $stat = stat($_REQUEST['path']);
  if (isset($stat['mtime'])) {
    $result['mtime'] = $stat['mtime'];
  }
}
$content = json_encode($result);
cachedEcho($content, "application/json");
```

Now when I right click on a torrent and select 'Fix Added Time' it does what I wanted but I get a popup (and a log)
Bad response from server: (200 [parserror, fix_addtime_request_action]) false

I suspect it has something to do with the response from action.php but I dont know how to fix it.
If somebody know please tell.
Either way I'm happy this works.

**Comment by Novik** [comment]
```
var base_path = theWebUI.torrents[plugin.uID].base_path;
```
1) base_path may be unknown if this torrent was not started as minimum once during this rtorrent session.
2) you should use encodeURIComponent for base_path (and rawurldecode in php code)

With this plugin everybody, who have access to ruTorrent, will may know - is concrete file or directory (any file, in any place) exists on your system or not. IMHO this is not good for security...

> Bad response from server: (200 [parserror, fix_addtime_request_action]) false

Check input/output parameters. Check web-server error log.
And of course, your plugin will not work with a master branch of ruTorrent.

**Reply by darko1979** [reply]
Thanks for the tips. I forgot to encode/decode path and because of that it did not work on all torrents, now it works :-)
As for security I do not care, it's for my personal use. I'm happy that it works.
As for bad response, I could not find a problem. Not sure why would ruTorrent report this. There is nothing in the error logs. Comparing request response from action.php to other plugins I don't see anything wrong, it returns proper 200 response, and at the end plugin receives this response and completes the task.
At the end it does not matter, I solved my problem :-)

Thanks for the help.
